xcube_clms/utils.py

```
@@ -640,3 +642,90 @@ def detect_format(url):
+                ds = xr.open_mfdataset(batch_paths, engine=engine, **kwargs)
+                LOG.info(f"Successfully opened batch of {len(batch_paths)} files")
+                return ds
+
+            except Exception as e:
+                LOG.warning(f"Attempt {attempt + 1} failed: {type(e).__name__}: {e}")
+
+                if attempt == max_retries - 1:
+                    LOG.error(f"All {max_retries} attempts failed for batch")
+                    raise e
+
+                delay = exponential_backoff(attempt)
+                LOG.info(f"Waiting {delay:.1f} seconds...")
+
+                time.sleep(delay)
+
+        return None
+
+    datasets = []
+    total_files = len(paths)
+
+    LOG.info(f"Processing {total_files} files in batches of {batch_size}")
+
+    for i in range(0, total_files, batch_size):
+        batch_end = min(i + batch_size, total_files)
+        batch_paths = paths[i:batch_end]
+        batch_num = (i // batch_size) + 1
+        total_batches = (total_files + batch_size - 1) // batch_size
+
+        LOG.info(
+            f"Processing batch {batch_num}/{total_batches} (files {i + 1}-{batch_end})"
+        )
+
+        try:
+            ds = open_batch_with_retry(batch_paths)
+            if ds is not None:
+                datasets.append(ds)
+
+            if i + batch_size < total_files:
+                time.sleep(1.5)
+
+        except Exception as e:
+            LOG.error(f"Failed to process batch {batch_num}: {str(e)}")
+            continue
+
+    if not datasets:
+        raise RuntimeError("No datasets could be opened successfully")
+
+    LOG.info(f"Successfully opened {len(datasets)} batches, combining...")
+
+    try:
+        combined_ds = xr.concat(datasets, dim="time")
+        LOG.info("Successfully combined all datasets")
+        return combined_ds
+    except Exception as e:
+        LOG.error(f"Failed to combine datasets: {str(e)}")
+        LOG.info("Returning list of datasets instead")
+        return datasets
```

This returns a list of datasets. If this then propagated to the user? 

This would break with the annotation of xcubes definition to return a `xr.Dataset`. Maybe raise an error instead.

---

Only reason I was doing that to avoid re-running everything again just because the concat failed. But if this breaks the annotation of xcubes definition, I will raise an error instead